Eagle-Rock-Analytics/historical-obs-platform
# QA/QC function: Flag frequent values

This function flags values if there are unusually frequent in the distribution. This occurs in two phases:
- Phase 1: data is scanned for all obs throughout the entire record, using a histogram where the bins represent 1degC/hPa. If any bin has 50+% greater number of observations that the total of the +/- 3 bins inclusive, it is highlighted for further review. Then each individual year in the observation record is assessed in the same manner, and if the highlighted bins still exhibit 50+% greater number of obs and 30+ obs (or 90+% and 20+ obs), all occurrences of that value in the record are flagged. If the year-by-year assessment reveals that it is not an unusually frequent value, the preliminary flag is dropped from review.
- Phase 2: the same process is repeated at the seasonal level, first all years of that season, then year by year. The thresholds are relaxed to account for fewer observations. 
    - Noteworthy: HadISD protocol is to do JF+D, with D being from the same year, **not** the previous year. This is especially problematic for the year by year check. When seasons are assessed year by year, the previous December is used instead of the current year's December. 
- Synergistic check: if air temp is flagged, then dew point is also flagged, and vice versa
- Histogram and timeseries* plots are produced and save to AWS
    - The timeseries plot function is in PR #103, and would need to be merged in first


----
**To test**
Run through the functions in the notebook, and check out the resulting plots in AWS. 
Example: ASOSAOWS_74718503144 does flag air temperature and therefore dewpoint

**Questions** to discuss in planning: 
- At present the histogram plots are produced for the entire time record for that variable, all obs all years all seasons. If a group is flagged it may be misleading. Do we configure these plots to do seasonal variations for specific years?
- This function is only run on air temp, dewpoint temp, and pressure according to HadISD. Do we include other variables?

**Updates**
- [x] update to new pipeline
- [x] update to pipeline modifications for index
- [x] check if setting month/year now necessary with index reset
- [x] update docstring
- [x] update era_qaqc_flag_meanings.csv 
- [x] update variables that proceed through check
- [x] update annual/seasonal flags + synergistic flag check
- [x] update figures for annual/seasonal checks
- [x] update to only use non-previously flagged data
- [x] diagnosing why bin check is now breaking
- [x] top level argument flag for radiation scheme for qaqc ease
